39 circshift hshift and vshift command for tables

raphael-sys opened the pull request:
### SUMMARY
Implements necessary changes for #39 

### IMPLEMENTATION
* Implementation: Implementation similar to description. Instead of 3 functions, there now are 4 functions: circshiftrow, circshiftcol, vectshiftrow, vectshiftcol. These functions cover all possible variations. The implementation of the column operation was straight forware, the row operations are currently implemented as wrappers of the columns, where the matrix is transposed and then handled by the col function. This works, but it has to be checked if this is efficient enough. Also there are currently too many int casts, as I was not sure why my negative values were not handled correctly.
* Implementation test: Manually tested each function with a 3x3 matrix and positive and negative values (plus 0) and also values larger than the dimensions of the matrix.

### DOCUMENTATION
* [x] ChangesLog updated
* [x] Code changes commented
* **Documentation articles:**
    * [ ] corresponding documentation articles updated
    * [ ] new documentation articles created
    * [x] not needed
* **Language files:**
    * [x] corresponding language files updated
    * [ ] not needed

----------------

### TESTS BY REVIEWERS
* [x] Added to the automatic SW tests
* [x] Tested manually

numeredev commented on kernel/core/maths/matfuncs.hpp:
```diff
@@ -3567,6 +3567,174 @@ static Matrix matrixFilter(const MatFuncData& funcData, const MatFuncErrorInfo&
+static Matrix vectShift(const MatFuncData& funcData, const MatFuncErrorInfo& errorInfo)
+{
+    // Check the dim argument
+    if (funcData.mVal < 0 || funcData.mVal > 1)
+        throw SyntaxError(SyntaxError::INVALID_AXIS, errorInfo.command, errorInfo.position);
```
Whether a check like this is actually necessary, is debatable. Especially considering, that the other parameter count is checked nowhere.

raphael-sys replied:
I do not know if it is really necessary, but could be helpful and also I would not know how to handle different inputs here. For the other parameter there are no invalid inputs, so no handling is required. 

Okay, complex numbers would be invalid input for the numbers of positions to shift, I forgot about that.

numeredev commented on kernel/core/maths/matfuncs.hpp:
```diff
@@ -3567,6 +3567,174 @@ static Matrix matrixFilter(const MatFuncData& funcData, const MatFuncErrorInfo&
+static Matrix circShift(const MatFuncData& funcData, const MatFuncErrorInfo& errorInfo)
+{
+    // Check the dim argument
+    if (funcData.mVal < 0 || funcData.mVal > 1)
+        throw SyntaxError(SyntaxError::INVALID_AXIS, errorInfo.command, errorInfo.position);
```
Same as in `vectShift`

numeredev commented on kernel/core/maths/matfuncs.hpp:
```diff
@@ -3567,6 +3567,174 @@ static Matrix matrixFilter(const MatFuncData& funcData, const MatFuncErrorInfo&
+static Matrix circShiftRow(const MatFuncData& funcData, const MatFuncErrorInfo& errorInfo)
+{
+    // Inputs: funcData.mat1 funcData.nVal
+    std::vector<mu::value_type> matData = funcData.mat1.data();
```
That's the first copy

numeredev commented on kernel/core/maths/matfuncs.hpp:
```diff
@@ -3567,6 +3567,174 @@ static Matrix matrixFilter(const MatFuncData& funcData, const MatFuncErrorInfo&
+    std::rotate(matData.begin(), matData.begin() + shift, matData.end());
+
+    Matrix _mResult = Matrix();
+    _mResult.assign(funcData.mat1.rows(), funcData.mat1.cols(), matData);
```
That's the second copy. Is that intended? You can get the internal array as a reference. Furthermore, if the passed matrix has been transposed from somewhere else, it is not working as expected. What about following this approach:
```C++
Matrix _mResult(funcData.mat1);
_mResult.extend(_mResult.rows(), _mResult.cols()); // force a transposition baking here

// Get the reference
std::vector<mu::value_type>& vData = _mResult.data();

// Call std::rotate() ...
return _mResult;
```
Then you won't need to call `extend()` twice in the opposite cases. Just transpose the matrix twice (the second one is the cheap call)

numeredev commented on kernel/core/maths/matfuncs.hpp:
```diff
@@ -3567,6 +3567,174 @@ static Matrix matrixFilter(const MatFuncData& funcData, const MatFuncErrorInfo&
+{
+    // Transpose the matrix (including restructuring the memory) to get a row wise representation of the data
+    Matrix _mResult = Matrix(funcData.mat1);
+    _mResult.transpose();
+    _mResult.extend(_mResult.rows(), _mResult.cols());
+
+    MatFuncData funcDataTransposed = MatFuncData(_mResult, funcData.nVal);
+
+    // Apply the circShiftRow to the transposed matrix
+    _mResult = circShiftRow(funcDataTransposed, errorInfo);
+
+    // Reverse the transposition
+    _mResult.transpose();
+    _mResult.extend(_mResult.rows(), _mResult.cols());
```
```suggestion
    _mResult.transpose();

    // Apply the circShiftRow to the transposed matrix
    _mResult = circShiftRow(MatFuncData(std::move(_mResult), funcData.nVal), errorInfo);

    // Reverse the transposition
    _mResult.transpose();
```

This should avoid uneccessary copies as the move constructor is called twice.

numeredev commented on kernel/core/maths/matfuncs.hpp:
```diff
@@ -3567,6 +3567,174 @@ static Matrix matrixFilter(const MatFuncData& funcData, const MatFuncErrorInfo&
+    std::vector<mu::value_type> matData = funcData.mat1.data();
+
+    // Circular shift downwards (negative values upwards)
+    // TODO: remove the unnecessary int casts
```
What about storing `matData.size()` as `int`. That will resolve many of the uncessary casts.

numeredev commented on kernel/core/maths/matfuncs.hpp:
```diff
@@ -3567,6 +3567,174 @@ static Matrix matrixFilter(const MatFuncData& funcData, const MatFuncErrorInfo&
+    for (size_t col = 0; col < funcData.mat1.cols(); col++)
+    {
+        // TODO: Remove the unnecessary int casts
+        int col_old = (int)((((int)col - (int)funcData.nVal) % (int)funcData.mat1.cols()) + (int)funcData.mat1.cols()) % (int)funcData.mat1.cols();
```
Same as before, store some of them as  `int`, that makes this line more readable

numeredev commented on kernel/core/maths/matfuncs.hpp:
```diff
@@ -3567,6 +3567,174 @@ static Matrix matrixFilter(const MatFuncData& funcData, const MatFuncErrorInfo&
+{
+    // Transpose the matrix (including restructuring the memory) to get a row wise representation of the data
+    Matrix _mResult = Matrix(funcData.mat1);
+    _mResult.transpose();
```
See the comment on `circShiftRow`

numeredev commented on kernel/core/maths/matfuncs.hpp:
```diff
@@ -3567,6 +3567,174 @@ static Matrix matrixFilter(const MatFuncData& funcData, const MatFuncErrorInfo&
+static Matrix vectShiftCol(const MatFuncData& funcData, const MatFuncErrorInfo& errorInfo)
+{
+    // Matrix data is a column-wise vector
+    std::vector<mu::value_type> matData = funcData.mat1.data();
```
Yeah, again an unecessary copy. See the comments on `circShiftCol` to improve the performance. Actually, it might be even simpler, if instead of using `std::copy` and doing all the transpositions, you want to write the data cell-wise directly to the target matrix. That might spare us some of the internal copies (which are only temporary in the end).

numeredev reviewed (changes requested):
I have the impression that by using `std::copy` and `std::rotate` you try to create elegant solution. However, they require a bunch of not needed copies, which can be avoided, if the elements are handled directly. Think about that.

raphael-sys commented:
I now cleaned up the int cast issues and removed two clearly unnecessary copies. Now the question remains whether this could be done more efficiently. In my opinion the row and col shift are good the way they are, direct addressing them would be complicated and probably not much faster. 
What do you think, should either the row shift or circ shift be solved via direct cell assignment?

numeredev reviewed (changes requested):
I think, using `std::rotate` is fine, but I'd recommend to implement the `vectshift()` functions by hand following my proposal: It avoids unecessary copies and it also solves the possible bug, which I just detected.

raphael-sys commented:
Ready for next review

numeredev commented on kernel/core/maths/matfuncs.hpp:
```diff
@@ -3684,8 +3685,9 @@ static Matrix vectShiftCol(const MatFuncData& funcData, const MatFuncErrorInfo&
-    Matrix _mResult = Matrix(funcData.mat1);
+    Matrix _mResult = funcData.mat1;
     _mResult.transpose();
+    // NOTE: This should not be necessary, but it does not work without it. Maybe the MatFuncData constructor does not copy the transposed state.
```
That's weird, have to investigate.

numeredev replied:
@raphael-sys Why do you think that the intended solution is not working? If I comment it out, it works as expected (ignore those `3x3` outputs):
![image](https://github.com/numere-org/NumeRe/assets/87755490/a4b15813-1ffc-450a-b27a-155b5bb1624d)

raphael-sys replied:
Cant reproduce the error now. So it probably is fine.

numeredev reviewed (approved):
PR accepted, will be merged
